digester.go

```diff
@@ -0,0 +1,253 @@
+	d.normalize(sql)
+	d.hasher.Write(d.buffer.Bytes())
+	d.buffer.Reset()
+	result = fmt.Sprintf("%x", d.hasher.Sum(nil))
+	d.hasher.Reset()
+	return
+}
+
+func (d *sqlDigester) doDigestText(sql string) (result string) {
+	d.normalize(sql)
+	result = string(d.buffer.Bytes())
+	d.lexer.reset("")
+	d.buffer.Reset()
+	return
+}
+
+const (
+	genericSymbol     = -1
+	genericSymbolList = -2
+)
+
+func (d *sqlDigester) normalize(sql string) {
+	d.lexer.reset(sql)
+	for {
+		tok, pos, lit := d.lexer.scan()
+		if tok == unicode.ReplacementChar && d.lexer.r.eof() {
+			break
+		}
+		if pos.Offset == len(sql) {
+			break
+		}
+		if tok == hintEnd {
+			d.reduceHint()
+			continue
+		}
+		currTok := token{tok, strings.ToLower(lit)}
+		if isLit(tok) {
+			currTok = d.reduceLit(currTok)
+		}
+		d.tokens = append(d.tokens, currTok)
+	}
+	d.lexer.reset("")
+	for i, token := range d.tokens {
+		d.buffer.WriteString(token.lit)
+		if i != len(d.tokens)-1 {
+			d.buffer.WriteRune(' ')
+		}
+	}
+	d.tokens = d.tokens[:0]
+}
+
+func (d *sqlDigester) reduceHint() {
+	for {
+		token := d.tokens.popBack(1)
+		if len(token) == 0 {
+			return
+		}
+		if token[0].tok == hintBegin {
+			return
+		}
+	}
+}
+
+func (d *sqlDigester) reduceLit(currTok token) token {
+	// "-x" or "+x" => "x"
+	if d.isPrefixByUnary(currTok.tok) {
+		d.tokens.popBack(1)
+	}
+
+	// "?, ?, ?, ?" => "..."
+	last2 := d.tokens.back(2)
+	if isGenericList(last2) {
+		d.tokens.popBack(2)
+		currTok.tok = genericSymbolList
+		currTok.lit = "..."
+		return currTok
+	}
+
+	// order by n => order by n
+	if currTok.tok == intLit {
+		last2 := d.tokens.back(2)
+		if isOrderBy(last2) {
+			return currTok
+		}
+	}
+
+	// 2 => ?
+	currTok.tok = genericSymbol
+	currTok.lit = "?"
+	return currTok
+}
+
+func (d *sqlDigester) isPrefixByUnary(currTok int) (isUnary bool) {
+	if !isNumLit(currTok) {
+		return
+	}
+	last := d.tokens.back(1)
+	if last == nil {
+		return
+	}
+	// a[0] != '-' and a[0] != '+'
+	if last[0].tok != 43 && last[0].tok != 45 {
+		return
+	}
+	last2 := d.tokens.back(2)
+	if last2 == nil {
+		isUnary = true
+		return
+	}
+	// '(-x' or ',-x' or ',+x' or '--x' or '+-x'
+	switch last2[0].tok {
+	case 40, 44, 43, 45, ge, is, le, eq, 60, 62:
+		isUnary = true
+	default:
+	}
+	// select -x or select +x
+	last2Lit := strings.ToLower(last2[0].lit)
+	if last2Lit == "select" {
+		isUnary = true
+	}
+	return
+}
+
+func isGenericList(last2 []token) (generic bool) {
+	if len(last2) < 2 {
+		return false
+	}
+	if !isComma(last2[1].tok) {
+		return false
+	}
+	switch last2[0].tok {
+	case genericSymbol, genericSymbolList:
+		generic = true
+	default:
+	}
+	return
+}
+
+func isOrderBy(last2 []token) (orderBy bool) {
+	if len(last2) < 2 {
+		return false
+	}
+	if last2[1].lit != "by" {
+		return false
+	}
+	orderBy = last2[0].lit == "order"
+	return
+}
+
+func isComma(tok int) (isComma bool) {
+	isComma = tok == 44
+	return
+}
+
+type token struct {
+	tok int
+	lit string
+}
+
+type tokenDeque []token
```

Comment:
`Queue` is enough?

Reply:
it will be push/pop in back and  foreach from front, maybe deque more suitable